### ISSUE TITLE
Serious regression / breaking change in activity SendHttpRequest of Elsa 2.9.0: response body looses double quotes (at least in some scenarios)!

### ISSUE THREAD

**Issue by Bergam64** [elsa-workflows/elsa-core#3385]
Hello Elsa Team,

I've recently upgraded from Elsa 2.8.2 to Elsa 2.9.1, and I've discovered a serious regression / breaking change in activity `SendHttpRequest`. After a few tests, it turns out that it has first been introduced in Elsa 2.9.0.

Here's my scenario (note: all my workflows are built by code):
- All my workflows call REST APIs on another server (using JSON formatting) via `SendHttpRequest`. Typical example:
  ```csharp
  .SendHttpRequest(request => request
      .WithUrl(() =>
      {
          var restUrl = new Uri(...);
  
          return restUrl;
      })
      .WithMethod(HttpMethods.Post)
      .WithContentType(MediaTypeNames.Application.Json)
      .WithContent(context =>
      {
          // Extract values from 'context'.
          
          var request = new SomeRequest
          {
              // Initialize request properties with extracted values.
          };
          string json = contentSerializer.Serialize(request);
  
          return json;
      })
      .WithReadContent(true)
  )
  .WithName("MyActivity")
  ```
- When asking creation of a business object on that server, the called RESP API always returns the id (i.e. the primary key, which is a `Guid`) of the created object. Example of raw response body returned by the remote server:
  ```json
  "28e49636-4ad1-4b68-80a7-ac774fecfd8d"
  ```
  (_notice the double quotes_).
- Then, after ensuring that the remote RESP API call was successful, the workflow retrieves that `Guid` ans stores it in a workflow variable with some code that boils down to this:
  ```csharp
  .SetVariable(
      "MyVariable",
      async context =>
      {
          object? rawResponseContent = await context
              .GetNamedActivityPropertyAsync<SendHttpRequest, object?>(
                  "MyActivity",
                  a => a.ResponseContent,
                  context.CancellationToken)
              .ConfigureAwait(false);

          if (rawResponseContent is null)
          {
              throw new InvalidOperationException("...");
          }

          string responseContent = (string)rawResponseContent;

          Guid createdObjectId = contentSerializer.Deserialize<Guid>(responseContent);

          return createdObjectId;
      })
  ```
- I've been doing this since Elsa 2.2.0 and this has worked fine up to Elsa 2.8.2: `responseContent` contains exactly this:
  ```
  "28e49636-4ad1-4b68-80a7-ac774fecfd8d"
  ```
  (notice that the double quotes are present, as expected), which allows successful deserialization to a `Guid`.
 - But with Elsa 2.9.0 (and 2.9.1), `responseContent` now contains this:
  ```
  28e49636-4ad1-4b68-80a7-ac774fecfd8d
  ```
  (**notice that the double quotes are magically gone**), which of course causes my code to miserably fail when it deserializes this to a `Guid`:
  ```
  Newtonsoft.Json.JsonReaderException:
  Input string '28e49636-4ad1-4b68-80a7-ac774fecfd8d' is not a valid number. Path '', line 1, position 36.
  ...
  ```

Bottom line: **all my workflows are now broken**, which means that Elsa 2.9.0 and 2.9.1 are show stoppers for me!

I don't have time to investigate this further, so I had no other choice but to revert to Elsa 2.8.2.

Please fix this !

**Comment by sfmskywalker** [comment]
Hi @Bergam64 , I'm sorry to see these changed behaviors are causing issues. Fortunately, it's easy to adapt to. All you need to do is configure the "old" behavior by configuring the `ResponseContentParserName` of the SendHttpRequest activity and set it to `"Plain Text"`. This will return the raw HTTP response body.

Example:

```csharp
.SendHttpRequest(request => request
    .WithUrl(() =>
    {
        var restUrl = new Uri(...);

        return restUrl;
    })
    .WithMethod(HttpMethods.Post)
    .WithContentType(MediaTypeNames.Application.Json)
    .WithContent(...)
    .WithReadContent(true)
    
     // Add this line.
    .Set(a => a.ResponseContentParserName, "Plain Text")
)
.WithName("MyActivity")
```

Alternatively, you can let the default parse be used and instead update your activity by replacing the following line:

```csharp
Guid createdObjectId = contentSerializer.Deserialize<Guid>(responseContent);
```

with this:

```csharp
Guid createdObjectId = Guid.Parse(responseContent);
```

**Comment by Bergam64** [comment]
Hello Sipke,

Thanks for your quick reply!

You seem to describe this as a breaking change but I've been unable to spot it in the release notes of [Elsa 2.9.0](https://github.com/elsa-workflows/elsa-core/releases/tag/2.9.0). Can you please ensure this is properly documented and where I can find it? Thanks in advance.

In the meantime, I need more info about this new default behavior. In my case, the response body is in fact a JSON with just a Guid in it and this new default behavior, for a reason that still eludes me, removes the surrounding double quotes. What is happening exactly, behind the scene?

Back on the issue topic, I find it questionable that _each_ use of activity `SendHttpRequest` must be updated to fix this issue. I already have more than 30 of them in my project. Is there any way to control this default behavior centrally?

**Comment by Bergam64** [comment]
Thanks for the analysis. BTW, impressive response time :thumbsup: !

So it turns out that it's a regression and not a breaking change that was missed to be documented in the release notes. Shall I update the issue title accordingly?

> Both parsers have a priority of 0, so the one selected will depend on the order in which these types are registered with the service container. I will push an update that changes the priority between these two.

I sincerely hope that this update will **also** fix the regression.
Will there be a 2.9.2 release or will it be only part of 2.10?

> If the `TypedHttpResponseContentReader` parser is selected, then the JSON string will be parsed into an `ExpandoObject`, unless the activity is configured to parse the response into a .NET type as specified via the `ResponseContentTargetType` property.

The fundamental problem I see when using the `ResponseContentTargetType` property is that Elsa assumes that **all** status codes (200, 400, 500, ...) will return a response body that can be always deserialized into the **same** .NET type, which seems a little overoptimistic IMHO and in my case is never true. That's why I manually deserialize the response body (with the approach explained on top) only in a branch of the `SendHttpRequest` activity dedicated to status code 200 (i.e. outcome "200"). For the other status codes (i.e. outcome "Unsupported Status Code"), I report the response body _as a string_ in a `WriteLine` activity then a I force the workflow to fail (with a custom `Fail` activity that throws an exception in its `OnExecute` method). I'm not aware of any other way to deal with this scenario.

I'd really like to know your opinion about this.

Thanks in advance!

**Comment by sfmskywalker** [comment]
No worries. The response time mileage varies from time to time, there will be others who might not be so impressed 😉 But I do do my best.

Yeah I would say it's an unintentional breaking change - feel free to update the title to something to that effect 👍🏻 

To fix this release, you'll need a way for the Text content parser to be selected, so that the response content is returned to you as a raw string.

There is currently a 2.9.2-rc2 release on NuGet. I'll try to do a 2.9.2-rc3 including this change.

Regarding the .NET type content parser, that's a good point. In case of an erroneous status code, it should probably not try and parse the response, but instead provide the body (if any) as a raw string value. Then it's up to the workflow designer to decide what to do with it, if anything. I'll give this some thought, but will probably out of scope for the next release candidate.

Thanks for the input!

**Comment by Bergam64** [comment]
Hello Sipke!

I finally found some time to upgrade to the latest Elsa (2.9.3) to assess the fix. First I was disappointed, as the issue didn't seem to have been fixed. It turned out that just upgrading was not enough. I _also had to guess_ that I needed to add this config entry in the Elsa server to the fix to kick in:
```json
  "Elsa": {
    "Server": {
      "DefaultContentParserName": "Plain Text"
    }
  }
```

It took me a while, as I _had to dig in the related commits_ to find that out!

The decision not to make this the default value inside Elsa is questionable, as this was about fixing a **regression**. At the very least, this should have been properly documented in the release notes, but unfortunately it wasn't.

As always, the biggest problem in Elsa is documentation. I hope you'll be able to increase the priority of this general issue in the near future, thereby making the life of all Elsa users **a lot easier**.

Anyway, thanks a lot for the fix.

**Comment by sfmskywalker** [comment]
Hi @Bergam64 , thank you for sharing your findings! I am glad you were able to figure it out in the end. You're 100% right, documentation + release notes need more attention and I'm committed to improving on this area by a lot.
